src/utils.rs

```diff
@@ -0,0 +1,83 @@
+// Compare two iterators to check if the needle is contained in the haystack.
+// Because iteration has to be restarted, the caller must provide two factory
+// functions that can skip to the provided offset.
+pub fn contains_kmp<F1, F2, I: Iterator<Item=char>>(haystack: F1, needle: F2) -> bool
+where F1 : Fn(usize) -> I, F2 : Fn(usize) -> I {
+    let mut left = haystack(0);
+    let mut right = needle(0);
+
+    // Consume the first needle character
+    let first = match right.next() {
+        None => return true,
+        Some(c) => c
+    };
+    
+    let mut next_possible_start = 0;
+    let mut i = 0;
+
+    loop {
+        // Consume left until a char matches the 1st needle char
+        let x = match left.next() {
+            None => return false,
+            Some(c) => c
+        };
+
+        i += 1;
+
+        if x != first {
+            continue;
+        }
+
+        let mut ix = 0;
+
+        loop {
+            let x = match left.next() {
```

Comment:
I think the order of getting a char of the left and the right should be reversed, so that we return true, if we reach the end of the needle at the same time as we reach the end of the haystack, in other words, if the needle is a suffix of the haystack.

The following test case fails:
```rust
    #[test]
    fn test_contains_finds_suffix() {
        let a = UniCase::unicode("ccbaab");
        let b = UniCase::unicode("baab");
        assert!(a.contains(&b));
    }
```